circuit_recognizer/circuit_recognizer/annotations.py

```
@@ -207,41 +222,51 @@ def from_component(cls, label, rect, name="image"):
     @staticmethod
     def from_dicts(dict_list):
         component_dict = {}
-        dict_list.sort(key=lambda data: 1 if "parent_id" in data else 0)
-        for data in dict_list:
-            if "parent_id" in data:
-                if data["parent_id"] not in component_dict:
-                    raise Exception(f'Could not find parent for component {data["id"]}')
-                parent = component_dict[data["parent_id"]]
-                component = PortAnnotation(**data)
-                parent.add_port(component)
-            else:
-                component = ComponentAnnotation(**data)
-                component_dict[component.id] = component
+        root_annotations = []
+
+        roots = filter(lambda data: "parent_id" not in data, dict_list)
+        for root in roots:
+            root_anno = ParentAnnotation(**root)
+            root_annotations.append(root_anno)
+            component_dict[root_anno.id] = root_anno
+
+        has_remaining_annotations = len(component_dict) != len(dict_list)
+        intermediate = filter(
+            lambda child: child.get("parent_id") in component_dict, dict_list
+        )
+        while has_remaining_annotations:
```

It is good that this removes the assumption that the annotations are ordered correctly. However, it might be nice to have a test for if the parent is actually missing from the list of annotations. (It should throw a meaningful error.)